specs/bls_verify.md

```diff
@@ -1,64 +1,117 @@
-### BLS Verification
+# BLS signature verification
 
 **Warning: This document is pending academic review and should not yet be considered secure.**
 
-See https://z.cash/blog/new-snark-curve/ for BLS-12-381 parameters. `q` is the field modulus.
+## Table of contents
+<!-- TOC -->
 
-We represent coordinates as defined in https://github.com/zkcrypto/pairing/tree/master/src/bls12_381/.
+- [BLS signature verification](#bls-signature-verification)
+    - [Table of contents](#table-of-contents)
+    - [Point representations](#point-representations)
+        - [G1 points](#g1-points)
+        - [G2 points](#g2-points)
+    - [Helpers](#helpers)
+        - [`hash_to_G2`](#hash_to_g2)
+        - [`modular_squareroot`](#modular_squareroot)
+    - [Signature verification](#signature-verification)
+        - [`bls_verify`](#bls_verify)
+        - [`bls_verify_multiple`](#bls_verify_multiple)
 
-Specifically, a point in G1 as a 384-bit integer `z`, which we decompose into:
+<!-- /TOC -->
 
-* `x = z % 2**381` (must be `< q`)
-* `highflag = z // 2**382`
-* `lowflag = (z % 2**382) // 2**381`
+## Curve 
 
-If `highflag == 3`, the point is the point at infinity and we require `lowflag = x = 0`. Otherwise, we require `highflag == 2`, in which case the point is `(x, y)` where `y` is the valid coordinate such that `(y * 2) // q == lowflag`.
+The BLS12-381 curve parameters are defined [here](https://z.cash/blog/new-snark-curve).
 
-We represent a point in G2 as a pair of 384-bit integers `(z1, z2)` that are each decomposed into `x1`, `highflag1`, `lowflag1`, `x2`, `highflag2`, `lowflag2` as above, where `x1` and `x2` must both be `< q`. We require `lowflag2 == highflag2 == 0`. If `highflag1 == 3`, the point is the point at infinity and we require `lowflag1 == x1 == x2 == 0`. Otherwise, we require `highflag == 2`, in which case the point is `(x1 * i + x2, y)` where `y` is the valid coordinate such that the imaginary part of `y` satisfies `(y_im * 2) // q == lowflag1`.
+## Point representations
 
-`BLSVerify(pubkey: uint384, msg: bytes32, sig: [uint384], domain: uint64)` is done as follows:
+We represent points in the groups G1 and G2 following [zkcrypto/pairing](https://github.com/zkcrypto/pairing/tree/master/src/bls12_381). We denote by `q` the field modulus and by `i` the imaginary unit.
 
-* Verify that `pubkey` is a valid G1 point and `sig` is a valid G2 point.
-* Convert `msg` to a G2 point using `hash_to_G2` defined below.
-* Do the pairing check: verify `e(pubkey, hash_to_G2(msg, domain)) == e(G1, sig)` (where `e` is the BLS pairing function)
+### G1 points
 
-Here is the `hash_to_G2` definition:
+A point in G1 is represented as a 384-bit integer `z` decomposed as a 381-bit integer and three 1-bit flags:
+
+* `x = z % 2**381`
+* `a_flag = (z % 2**382) // 2**381`
+* `b_flag = (z % 2**383) // 2**382`
+* `c_flag = (z % 2**384) // 2**383`
+
```

Comment:
Would be nice to 
- emphasize flags are prefixed. It takes a reader a while to figure out the relative position of bits from bullet points. `z = |c_flag|b_flag|a_flag|...x...|`
- and intepret the intention of these flags. Say, b_flag indicates if it is infinity, c_flag is dummy, while a_flag identifies which of two possible y this point has.

Reply:
> emphasize flags are prefixed. It takes a reader a while to figure out the relative position of bits from bullet points. `z = |c_flag|b_flag|a_flag|...x...|`

Done :)

> and intepret the intention of these flags

The problem is that these flags have different meanings in G1 and G2.